ts/haskell/mod.test.ts

```diff
@@ -1,8 +1,60 @@
 import { describe, it } from "https://deno.land/std@0.206.0/testing/bdd.ts";
+import outdent from "https://deno.land/x/outdent@v0.8.0/mod.ts";
 import * as haskell from "./mod.ts";
 import { assertStdout, assertSuccess, runExecutable } from "../testUtils.ts";
 
 describe("mkHaskellProject", () => {
+  describe("withVersions", () => {
+    it("allows specifying alternative versions from Hackage", () => {
+      const path = Deno.makeTempDirSync();
+      Deno.writeTextFileSync(
+        `${path}/project.cabal`,
+        outdent`
+          name: project
+          version: 0.0.1
+          executable foo
+            main-is: Main.hs
+            build-depends: base, string-conversions
+            default-language: Haskell2010
+        `,
+      );
+      Deno.writeTextFileSync(
+        `${path}/Main.hs`,
+        outdent`
+          {-# LANGUAGE CPP #-}
+
+          #if MIN_VERSION_string_conversions(0,4,0)
+          main = putStrLn ">= 0.4"
+          #else
+          main = putStrLn "< 0.4"
+          #endif
+        `,
+      );
+      let project = haskell.mkHaskellProject({
+        description: "",
+        compiler: "ghc94",
+        src: ".",
+        executables: ["foo"],
+        overrideDependencies: {
+          "string-conversions": "0.3.0.3",
+        },
+      });
+      let output = assertSuccess(runExecutable(project.foo, { cwd: path }));
+      assertStdout(output, "< 0.4\n");
+      project = haskell.mkHaskellProject({
+        description: "",
+        compiler: "ghc94",
+        src: ".",
+        executables: ["foo"],
+        overrideDependencies: {
+          "string-conversions": "0.4.0.1",
+        },
+      });
+      output = assertSuccess(runExecutable(project.foo, { cwd: path }));
+      assertStdout(output, ">= 0.4\n");
+    });
+  });
```

Comment:
I pushed fbe02289 as a review commit. Personally I find it easier to have the two project examples clearly separated. And re-using variables is also ok, although I care less about that.